### DISCUSSION TITLE
Creating a class primitive?

### DISCUSSION THREAD

**Discussion by sanders3d** [Autodesk/maya-usd#3262] [General]
Hello, I'm wondering if it's possible with the maya usd plugin to create a "class" prim as it's stated here in this doc.
https://github.com/usd-wg/assets/blob/main/docs/asset-structure-guidelines.md

I can create them in Houdini but I cant figure out how to create it with the stage manager or any other way in maya. Are we missing a "class" prim? I can't see any other prim that I can adjust to become a "class"

example of a simple asset that has a "rubber_duck" class.

```
#usda 1.0
(
    endTimeCode = 1
    framesPerSecond = 24
    metersPerUnit = 1
    startTimeCode = 1
    timeCodesPerSecond = 24
    upAxis = "Y"
)

class "__class__"
{
    class "rubber_duck"
    {
    }
}

def "rubber_duck" (
    prepend payload = @./payloads/payload.usda@</rubber_duck>
)
{
    custom float3[] extentsHint
    float3[] extentsHint.timeSamples = {
        1: [(-0.7087299, -0.016021485, -0.48650336), (0.6047075, 0.16225857, 0.4517015)],
    }

}
```
Any suggestion  would be helpful

**Comment by sanders3d** [comment]
So I have managed to create a script that actually creates two class primitives.

```
import maya.cmds as cmds
import mayaUsd
import mayaUsd_createStageWithNewLayer
from pxr import Usd, UsdGeom, Sdf

def SaveStageTest():
    stage_shape = mayaUsd_createStageWithNewLayer.createStageWithNewLayer()
    tmp_root = cmds.listRelatives(stage_shape, parent=True)[0]
    
    stage_root = cmds.rename(tmp_root, "rubber_duck")
    stage = mayaUsd.ufe.getStage(cmds.listRelatives(stage_root, shapes=True, fullPath=True)[0])
    stage_layer = stage.GetRootLayer()
    
    class_root_prim = stage.DefinePrim("/__class__")
    class_name_prim = stage.DefinePrim("/__class__/rubberduck")    
    class_root_prim.GetPrim().SetSpecifier(Sdf.SpecifierClass)
    class_name_prim.GetPrim().SetSpecifier(Sdf.SpecifierClass)
        
    asset_prim = stage.DefinePrim('/ASSET', "Xform")
    stage_layer.Export("/tmp/test.usda")
        
SaveStageTest() 
```   
And the USD file looks like this

```
#usda 1.0

class "__class__"
{
    class "rubberduck"
    {
    }
}

def Xform "ASSET"
{
}

```
The problem though is that they user can't see any of the __class__ or "rubberduck" in the layer editor. So it's hard for them to know if everything was created correctly without exporting and looking inside of the USD file.

Would be very helpful if we can view these items in the layer editor.

**Reply by womanyoyoyo-adsk** [reply]
Hi @sanders3d, there is no way to create class prims right now in MayaUSD's UI. We don't display them in the Outliner either. When you say it would be helpful to see them in the Layer Editor, do you mean Maya's Outliner (the tree of prims)? Or how would you expect to see them in Layer Editor?

**Reply by sanders3d** [reply]
yes I meant the outliner. When you turn on classes in usdview you can see them, and it's and important tool to have especially when you export collections of objects that you use in a scene assembly. So this could be an option in the stage to view or not view ( just like you have in usdview ). 
And when you export you should be able to add a `prepend inherit` to that class.

![image](https://github.com/Autodesk/maya-usd/assets/44543707/40afa406-9941-4d65-87ba-3b80eaa19038)

I hope that makes sense.
regards
stefan

**Reply by womanyoyoyo-adsk** [reply]
Displaying class prims in the Outliner makes sense. Can you clarify what you mean by export? Right now we export Maya data to USD data. I'd like to understand when you'd expect to prepend an inherit to the class prim. Thank you!

**Reply by sanders3d** [reply]
I say export since we need to control where and how the layer is saved. So we never use the interface button ( which I wish wasn't there )

fake code example
```
out = "/tmp/test.usda"
root_layer = stage.GetRootLayer()
root_layer.identifier = out
root_layer.Save()
stage.Export(out)
```

and here is a few snippets on how I add prepend

```
    def create_class(self):      
        class_root_prim = self.stage.DefinePrim("/__class__")
        self.class_name_prim = self.stage.DefinePrim(f"/__class__/{self.name}")
        class_root_prim.GetPrim().SetSpecifier(Sdf.SpecifierClass)
        self.class_name_prim.GetPrim().SetSpecifier(Sdf.SpecifierClass)

    def create_default_prim(self):
        self.default_prim = UsdGeom.Xform.Define(self.stage, f"/{self.name}")
        self.stage.SetDefaultPrim(self.default_prim.GetPrim())

    def add_inherit_class(self):
        inherits = self.default_prim.GetPrim().GetInherits()
        inherits.AddInherit(self.class_name_prim.GetPath()) 
```

regards
stefan

**Reply by womanyoyoyo-adsk** [reply]
Thank you.
